test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java

```diff
@@ -25,7 +25,7 @@
  * @test
  * @bug 8303742
  * @summary CompletableFuture.orTimeout can leak memory if completed exceptionally
- * @run junit/othervm -Xmx128m CompletableFutureOrTimeoutExceptionallyTest
+ * @run junit/othervm/timeout=1000 -Xmx128m CompletableFutureOrTimeoutExceptionallyTest
```

Comment:
If there isn't any value running this test with a debug build then you could have it skip (`@requires vm.debug == true`) to save resources.

Reply:
@jaikiran Having a long timeout doesn't seem like a problem given that it just needs enough time to run through the iterations (i.e. that's the max duration of the test before giving up).

@AlanBateman That's a great observation—do you see anything about this which would impact whether it makes sense to run on top of a debug-build (I don't see any). If it doesn't need to run on a debug build then I'll add the requirement you suggested. Let me know.

Reply:
@AlanBateman Wouldn't that be ´@requires vm.debug == false` ?

Reply:
Hello Viktor,

> @jaikiran Having a long timeout doesn't seem like a problem given that it just needs enough time to run through the iterations (i.e. that's the max duration of the test before giving up).

Having a extremely long timeout of T1 secods for a test which one can (reasonably) expect to finish in (T1 - X) seconds can mean that when it genuinely times out, then it holds on to the limited shared resources (like the host machine) for those X seconds longer instead of potentially letting other tests run during that period. So I think it's always better to have a reasonable timeout instead of an extremely large one - a value past which you know that if the test is still running then it's surely be a sign that it should no longer continue to run.
Typically the timeout for such tests is decided by running the test against the hosts/environment where it failed and gathering data to see how long it usually takes to finish successfully on those and then adding some extra seconds to it.

Reply:
@jaikiran The downside is that it is hard to judge how long it will take for machine X to execute Y instructions. :/

Reply:
1. I don't know.
2. It's hard to come up with a one-size-fits-all-heap-size given that differences in pointersizes, object alignment, etc can skew the need.
3. I initially set the threshold (iteration number) at about 4x what I needed locally to make it fail, just to have margin in case memory-usage-per-object would vary with the different setups being tested. 


The other alternative I see would be to reach into the implementation of CompletableFuture's `Delayer`'s `ScheduledThreadPoolExecutor delayer` and make sure that it's `getQueue()` eventually goes empty.

Reply:
> The other alternative I see would be to reach into the implementation of CompletableFuture's `Delayer`'s `ScheduledThreadPoolExecutor delayer` and make sure that it's `getQueue()` eventually goes empty.

From what I've seen, JDK prefers blackbox tests to whitebox tests. Even though the latter might conserve resources better and are easier to implement, they are problematic in the long run: they become an obstacle if the system under tests is modified.

Reply:
@pavelrappo I agree with that. I have the same experience. With that said, I have implemented a whitebox test which reaches in an monitors the task queue of the Delayer. @AlanBateman & @jaikiran any preference here?

Reply:
@jaikiran Yes, see the new commit I pushed which illustrates the whitebox solution (which will also run dramatically faster than the old one) which relies on the implementation detail on how the timeouts are scheduled.

@AlanBateman @pavelrappo Does the new approach seem acceptable? 🤔

Reply:
> @AlanBateman @pavelrappo Does the new approach seem acceptable? 

I don't have a strong opinion on this. The advantage is that the test is probably sub-second as it just needs one CF to complete exceptionally.  The disadvantage is that it the test would need to be updated if the implementation changes, but it's not hard.

So good with the approach if you want. As Jai mentioned, you can use @modules to open java.base/java.util.concurrent for deep reflectively. I would probably trim the overly long lines as they currently wrap when looking at the diffs.

Reply:
While the approach seems okay, it might be for the JSR 166 folk to make the final decision on it.

(Looking at Thread.sleep(). Sigh: unfortunately, BlockingQueue does not provide an option to examine its head that blocks or times out.)

Reply:
@pavelrappo `peek(timeout, unit)` would be nice :)

@AlanBateman Sounds good, I'll change to use @modules :+1:

Reply:
@AlanBateman @jaikiran @pavelrappo I've updated the code to use an "@modules" (thanks for the hint!) and reformatted the code slightly. I think we should be good to go here.